File: clase_15/code/t1_nolock.c
#include <stdio.h>
#include <stdlib.h>
#include <pthread.h>
#include <sched.h>

/*
Ejemplo SIN lock: demuestra la race condition (condicion de carrera)
al incrementar un contador compartido desde dos hilos concurrentes.

Comparar con t1_lock.c para ver el efecto de agregar un mutex.

1. To compile:
   gcc -O0 -o t1_nolock t1_nolock.c -lpthread -Wall

2. To run:
   ./t1_nolock <MAX_VALUE>

   Ejemplo:
   ./t1_nolock 1000

3. Salida esperada (con race condition):

   main: begin [counter = 0]
   A: done
   B: done
   main: done
   [counter: 1001]      <- valor INCORRECTO (deberia ser 2000)
   [should:  2000]

   Nota: el valor del counter sera casi siempre menor que (max * 2)
   y puede variar entre ejecuciones.

4. Por que ocurre la race condition?

   La linea  counter = counter + 1  genera 3 instrucciones en ensamblador:

      mov 0x<addr>, %eax     <- (1) lee counter desde memoria al registro
      add $0x1,     %eax     <- (2) incrementa el registro
      mov %eax,     0x<addr> <- (3) escribe el resultado en memoria

   Si el scheduler interrumpe un hilo entre (1) y (3) y el otro hilo
   ejecuta sus 3 pasos completos, la escritura del segundo hilo se
   pierde cuando el primer hilo ejecuta (3) con su valor desactualizado.

   sched_yield() fuerza ese cambio de contexto para que la race condition
   sea siempre observable. En produccion ocurre igual pero de forma
   no determinista.
*/

int max;
int counter = 0; // shared global variable - SIN proteccion

void *mythread(void *arg) {
    char *letter = arg;
    int i; // stack (private per thread)
    for (i = 0; i < max; i++) {
        int tmp = counter;  // (1) lee counter en variable local (registro)
        sched_yield();      // fuerza cambio de contexto: el otro hilo corre aqui
        tmp = tmp + 1;      // (2) incrementa la copia local
        counter = tmp;      // (3) escribe -- puede pisar el valor del otro hilo
    }
    printf("%s: done\n", letter);
    return NULL;
}

int main(int argc, char *argv[]) {
    if (argc != 2) {
        fprintf(stderr, "usage: t1_nolock <loopcount>\n");
        exit(1);
    }
    max = atoi(argv[1]);

    pthread_t p1, p2;
    printf("main: begin [counter = %d]\n", counter);
    pthread_create(&p1, NULL, mythread, "A");
    pthread_create(&p2, NULL, mythread, "B");
    pthread_join(p1, NULL);
    pthread_join(p2, NULL);
    printf("main: done\n");
    printf("[counter: %d]\n[should:  %d]\n", counter, max * 2);
    return 0;
}

File: clase_15/code/t1_nolock.py
import threading
import sys
import time

# Necesario usar un objeto para que sea mutable y compartido entre hilos
class Counter:
    def __init__(self):
        self.value = 0

counter = Counter()
max_count = 0  # renombrado para no pisar el built-in max() de Python

def mythread(arg):
    print(f"{arg}: begin")
    for i in range(max_count):
        # SIN lock: multiples hilos pueden leer, incrementar y
        # escribir counter.value al mismo tiempo.
        #
        # Equivalente en Python a las 3 instrucciones de ensamblador:
        #   mov 0x<addr>, %eax     <- (1) lee counter desde memoria
        #   add $0x1,     %eax     <- (2) incrementa el registro
        #   mov %eax,     0x<addr> <- (3) escribe el resultado en memoria
        #
        # time.sleep(0) libera el GIL entre (1) y (3), permitiendo que
        # el otro hilo ejecute sus 3 pasos completos y pise el valor.
        # Es el equivalente Python de sched_yield() en C.
        #
        # Nota: el GIL de CPython reduce (pero no elimina) las race
        # conditions. Sin time.sleep(0) pueden no observarse con 2 hilos.
        tmp = counter.value   # (1) lee
        time.sleep(0)         # libera el GIL — el otro hilo corre aqui
        tmp = tmp + 1         # (2) incrementa la copia local
        counter.value = tmp   # (3) escribe -- puede pisar el valor del otro hilo
    print(f"{arg}: done")

if __name__ == "__main__":
    if len(sys.argv) != 2:
        print("usage: python3 t1_nolock.py <loopcount>")
        sys.exit(1)

    max_count = int(sys.argv[1])
    print("main: begin")
    print(f"main: begin [counter = {counter.value}]")

    p1 = threading.Thread(target=mythread, args=('A',))
    p2 = threading.Thread(target=mythread, args=('B',))
    p1.start()
    p2.start()
    p1.join()
    p2.join()

    print(f"main: done")
    print(f"[counter: {counter.value}]")
    print(f"[should:  {max_count * 2}]")